runtime/compiler/optimizer/J9ValuePropagation.cpp

```diff
@@ -1175,3 +1176,167 @@ J9::ValuePropagation::getParmValues()
+   {
+   // Due to VirtualGuardHeadMerger, the taken side of a virtual guard may have more than one predecessors,
+   // each containing a virtual guard that branches to the taken side. It's sufficient to look at only
+   // one predecessor.
+   TR::Node* predLastRealNode = block->getPredecessors().size() > 0 ?
+                                block->getPredecessors().front()->getFrom()->asBlock()->getLastRealTreeTop()->getNode() :
+                                NULL;
+   if (predLastRealNode
+       && predLastRealNode->isTheVirtualGuardForAGuardedInlinedCall()
+       && predLastRealNode->getBranchDestination()->getEnclosingBlock() == block)
+      return true;
+
+   return false;
+   }
+
+static bool skipFinalFieldFoldingInBlock(TR::Block* block)
+   {
+   if (block->isCold() ||
+       isOSRRelatedBlock(block) ||
+       isTakenSideOfAVirtualGuard(block))
+      return true;
+
+   return false;
+   }
+
+static bool isStaticFinalFieldWorthFolding(TR::Compilation* comp, TR_OpaqueClassBlock* declaringClass, char* fieldSignature, int32_t fieldSigLength)
+   {
+   if (comp->getMethodSymbol()->hasMethodHandleInvokes()
+       && !TR::Compiler->cls.classHasIllegalStaticFinalFieldModification(declaringClass))
+      {
+      if (fieldSigLength == 28 && !strncmp(fieldSignature, "Ljava/lang/invoke/VarHandle;", 28))
+         return true;
+      }
+
+   return false;
+   }
+
+
+// Do not add fear point in a frame that doesn't support OSR
+static bool safeToAddFearPointIn(TR::Compilation* comp, int32_t callerIndex)
+   {
+   TR::ResolvedMethodSymbol *method = callerIndex == -1 ?
+      comp->getJittedMethodSymbol() : comp->getInlinedResolvedMethodSymbol(callerIndex);
+
+   return !method->cannotAttemptOSRDuring(callerIndex, comp, false);
+   }
+
+bool J9::ValuePropagation::transformDirectLoad(TR::Node* node)
+   {
+   // Allow OMR to fold reliable static final field first
+   if (OMR::ValuePropagation::transformDirectLoad(node))
+      return true;
+
+   if (node->isLoadOfStaticFinalField() &&
+       tryFoldStaticFinalFieldAt(_curTree, node))
+      {
+      return true;
+      }
+
+   return false;
+   }
+
+/** \brief
+ *     Try to fold static final field
+ *
+ *  \param tree
+ *     The tree with the load of static final field.
+ *
+ *  \param fieldNode
+ *     The node which is a load of a static final field.
+ */
+bool J9::ValuePropagation::tryFoldStaticFinalFieldAt(TR::TreeTop* tree, TR::Node* fieldNode)
+   {
+   TR_ASSERT(fieldNode->isLoadOfStaticFinalField(), "Node n%dn %p has to be a load of a static final field", fieldNode->getGlobalIndex(), fieldNode);
+
+   if (comp()->getOption(TR_DisableGuardedStaticFinalFieldFolding))
+      {
+      return false;
+      }
+
+   if (!comp()->supportsInduceOSR()
+       || !comp()->isOSRTransitionTarget(TR::postExecutionOSR)
+       || !comp()->getOSRMode() == TR::voluntaryOSR)
+      {
+      return false;
+      }
+
+   if (fieldNode->getByteCodeInfo().doNotProfile()
```

Comment:
hmm - I'm not sure I like this use of doNotProfile - whether a static final field read is foldable is not related to whether the field read node came from the original program representation so this would seem an overload of the meaning of the flag likely to confuses other opts